src/lightning/fabric/utilities/optimizer.py

```diff
@@ -30,5 +29,11 @@ def _optimizers_to_device(optimizers: Iterable[Optimizer], device: _DEVICE) -> N
 
 def _optimizer_to_device(optimizer: Optimizer, device: _DEVICE) -> None:
     """Moves the state of a single optimizer to the device."""
-    for p, v in optimizer.state.items():
-        optimizer.state[p] = apply_to_collection(v, Tensor, move_data_to_device, device, allow_frozen=True)
+    for _, v in optimizer.state.items():
+        if not isinstance(v, MutableMapping):
+            continue
```

Comment:
In this case, maybe we should still do `optimizer.state[p] = apply_to_collection(v, Tensor, move_data_to_device, device, allow_frozen=True)`, for custom optimizers that don't use dicts to store the state?